src/stage1/bigint.cpp

```diff
@@ -468,6 +468,76 @@ void bigint_min(BigInt* dest, const BigInt *op1, const BigInt *op2) {
     }
 }
 
+/// clamps op within bit_count/signedness boundaries
+void bigint_clamp_by_bitcount(BigInt* dest, uint32_t bit_count, bool is_signed) {
+    bool is_negative = dest->is_negative;
+    // set to false so that bigint_bits_needed() won't report 64+ bits needed for small negative numbers
+    dest->is_negative = false;
+    
+    // compute the number of bits required to store the value, and use that 
+    // to decide whether to clamp the result
+    size_t bits_needed = bigint_bits_needed(dest) + is_negative;
```

Comment:
For i32 you have [-2^31..2^31-1] vs u32 [0..2^32] and I dont see an explanation how this is handled.
Adding a few words would help to understand follow-up code.